src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetails.cs

```diff
@@ -25,9 +27,49 @@ public abstract class RethrowToPreserveStackDetailsAnalyzer : DiagnosticAnalyzer
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
-        protected static Diagnostic CreateDiagnostic(SyntaxNode node)
+        public override void Initialize(AnalysisContext analysisContext)
         {
-            return node.CreateDiagnostic(Rule);
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterOperationAction(context =>
+            {
+                var throwOperation = (IThrowOperation)context.Operation;
+
+                if (throwOperation.Exception is not ILocalReferenceOperation localReference)
+                {
+                    return;
+                }
+
+                IOperation ancestor = throwOperation;
+                while (ancestor != null &&
+                    ancestor.Kind != OperationKind.AnonymousFunction &&
+                    ancestor.Kind != OperationKind.LocalFunction)
+                {
+                    if (ancestor.Kind == OperationKind.CatchClause &&
+                        ancestor is ICatchClauseOperation catchClause)
```

Comment:
Is the first check necessary?
I think ***every*** ICatchClauseOperation should have OperationKind.CatchClause and vice-versa.

```suggestion
                    if (ancestor is ICatchClauseOperation catchClause)
```

Reply:
@Youssef1313 I am doing the first check for a fast bail-out, I don't know if that's something we care about. Some analyzers do some don't.

@mavasani any opinion on this?

Reply:
Seems fine either way to me.

Reply:
If it's known that the second check is slower, then keep both. I wasn't aware that the first check is faster than the second.